From 754f8cb72b42a3a6100d2bbb1cb885361a7310dd Mon Sep 17 00:00:00 2001
From: Vivek Goyal <vgoyal@redhat.com>
Date: Fri, 1 Jul 2016 16:34:29 -0400
Subject: [PATCH] ovl: do not require mounter to have MAY_WRITE on lower
References: bsc#1010158
Git-commit: 754f8cb72b42a3a6100d2bbb1cb885361a7310dd
Patch-mainline: 4.8

Now we have two levels of checks in ovl_permission(). overlay inode
is checked with the creds of task while underlying inode is checked
with the creds of mounter.

Looks like mounter does not have to have WRITE access to files on lower/.
So remove the MAY_WRITE from access mask for checks on underlying
lower inode.

This means task should still have the MAY_WRITE permission on lower
inode and mounter is not required to have MAY_WRITE.

It also solves the problem of read only NFS mounts being used as lower.
If __inode_permission(lower_inode, MAY_WRITE) is called on read only
NFS, it fails. By resetting MAY_WRITE, check succeeds and case of
read only NFS shold work with overlay without having to specify any
special mount options (default permission).

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>
Signed-off-by: Miklos Szeredi <mszeredi@redhat.com>
Signed-off-by: Fabian Vogt <fvogt@suse.com>
---
 fs/overlayfs/inode.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/fs/overlayfs/inode.c
+++ b/fs/overlayfs/inode.c
@@ -166,6 +166,11 @@ int ovl_permission(struct inode *inode,
 			goto out_dput;
 	}
 
+	if (!is_upper && !special_file(realinode->i_mode) && mask & MAY_WRITE) {
+		mask &= ~(MAY_WRITE | MAY_APPEND);
+		/* Make sure mounter can read file for copy up later */
+		mask |= MAY_READ;
+	}
 	err = __inode_permission(realinode, mask);
 out_dput:
 	dput(alias);
